staging/src/k8s.io/apiserver/pkg/endpoints/filters/audit.go

```diff
@@ -101,62 +119,200 @@ func WithAudit(handler http.Handler, requestContextMapper request.RequestContext
-			asgroups = auditStringSlice(requestedGroups)
+
+		if err := updateEventFromRequest(ev, req, ctx); err != nil {
+			glog.Errorf("Failed to complete audit event from request: %v", err)
+			responsewriters.InternalError(w, req, errors.New("failed to complete audit event from request"))
+			return
 		}
-		namespace := attribs.GetNamespace()
-		if len(namespace) == 0 {
-			namespace = "<none>"
+
+		// we have two code path:
+		//
+		// a) short running request with one event:
+		//
+		//    Wait synchronously for the handler to finish and send a complete event.
+		//
+		// b) long running requests with two events:
+		//
+		//    Spawn a go routine waiting for the status code, synchronously send a first
+		//    request audit event and then give back control to the handler.
+		//    Then continue with a).
+		var firstEventSentCh chan struct{}
+		longRunning := false
+		if longRunningCheck != nil {
+			ri, _ := request.RequestInfoFrom(ctx)
+			if longRunning = longRunningCheck(req, ri); longRunning {
+				firstEventSentCh = make(chan struct{}, 0)
+				defer func() {
+					close(firstEventSentCh)
+				}()
+			}
 		}
-		id := uuid.NewRandom().String()
+		respWriter, codeCh := decorateResponseWriter(w, firstEventSentCh)
 
-		line := fmt.Sprintf("%s AUDIT: id=%q ip=%q method=%q user=%q groups=%q as=%q asgroups=%q namespace=%q uri=%q\n",
-			time.Now().Format(time.RFC3339Nano), id, utilnet.GetClientIP(req), req.Method, username, groups, asuser, asgroups, namespace, req.URL)
-		if _, err := fmt.Fprint(out, line); err != nil {
-			glog.Errorf("Unable to write audit log: %s, the error is: %v", line, err)
+		// for long-running requests send a first event when the status code arrives
+		if longRunning {
+			go func() {
+				defer close(firstEventSentCh)
+
+				code, ok := <-codeCh
+				if !ok {
+					// closed channel
+					return
+				}
+
+				// here the handler is waiting for us synchronously
+
+				if err := backend.LogAuditEvent(ev); err != nil {
+					glog.Errorf(err.Error())
+				}
+
+				// set the code in the second event.
+				if ev.ResponseStatus == nil {
+					ev.ResponseStatus = &metav1.Status{}
+				}
+				ev.ResponseStatus.Code = int32(code)
+			}()
 		}
-		respWriter := decorateResponseWriter(w, out, id)
+
+		// send audit event when we leave this func, either via a panic or cleanly. In the case of long
+		// running requests, this will be the second audit event.
+		defer func() {
+			if longRunning {
+				// synchronize with the go routine. If a status code has been written, the
+				// go routine above will get the first chance to send an audit event for it.
+				close(codeCh)
+				<-firstEventSentCh
+			}
+
+			if r := recover(); r != nil {
+				ev.ResponseStatus = &metav1.Status{
+					Code: http.StatusInternalServerError,
+				}
+				if err := backend.LogAuditEvent(ev); err != nil {
+					glog.Errorf(err.Error())
+				}
+				panic(r)
+			}
+
+			if ev.ResponseStatus == nil {
+				ev.ResponseStatus = &metav1.Status{
```

Comment:
I'm not sure filling in Status.Status should be considered making up data. It should just be a summary of the response code. Is there anywhere that it's not a direct mapping from the code?

Of course we could easily do this on the receiving side of the audit event, but I thought it would be cleaner to always populate it...